File: wasm/smoke.cpp
// WASM feasibility smoke — compile a JSON graph definition + run a
// trivial graph entirely inside the engine. No HTTP, no LLM, no
// network. If this builds + runs under em++, Phase 1 is feasible.

#include <neograph/graph/engine.h>
#include <neograph/graph/state.h>
#include <neograph/graph/node.h>
#include <neograph/graph/types.h>
#include <neograph/json.h>

#include <cstdio>
#include <memory>
#include <string>
#include <vector>

using namespace neograph;
using namespace neograph::graph;

class DoubleNode : public GraphNode {
public:
    explicit DoubleNode(std::string name) : name_(std::move(name)) {}
    std::string get_name() const override { return name_; }
    asio::awaitable<NodeOutput> run(NodeInput in) override {
        auto seed = in.state.get("seed");
        int v = seed.is_number() ? seed.get<int>() : 0;
        NodeOutput out;
        out.writes.push_back(ChannelWrite{"doubled", json(v * 2)});
        co_return out;
    }
private:
    std::string name_;
};

int main() {
    NodeFactory::instance().register_type(
        "doubler",
        [](const std::string& name, const json&, const NodeContext&)
            -> std::unique_ptr<GraphNode> {
            return std::make_unique<DoubleNode>(name);
        });

    json defn = {
        {"name", "wasm_smoke"},
        {"channels", {
            {"seed",    {{"reducer", "overwrite"}}},
            {"doubled", {{"reducer", "overwrite"}}},
        }},
        {"nodes",    {{"d", {{"type", "doubler"}}}}},
        {"edges", json::array({
            {{"from", "__start__"}, {"to", "d"}},
            {{"from", "d"},         {"to", "__end__"}},
        })},
    };

    auto engine = GraphEngine::compile(defn, NodeContext());
    engine->set_checkpoint_store(std::make_shared<InMemoryCheckpointStore>());

    RunConfig cfg;
    cfg.thread_id = "wasm-1";
    cfg.input     = json{{"seed", 21}};
    cfg.max_steps = 5;

    auto result = engine->run(cfg);
    auto out = result.output["channels"]["doubled"]["value"];
    std::printf("doubled = %d\n", out.get<int>());
    std::printf("trace = ");
    for (const auto& step : result.execution_trace) std::printf("%s ", step.c_str());
    std::printf("\n");
    return out.get<int>() == 42 ? 0 : 1;
}

File: wasm/smoke.mjs
import assert from 'node:assert/strict'
import path from 'node:path'
import { pathToFileURL } from 'node:url'

const modulePath = path.resolve(process.argv[2] ?? 'build/wasm-release/neograph.js')
const createNeoGraphModule = (await import(pathToFileURL(modulePath).href)).default
const runtime = await createNeoGraphModule()

assert.match(runtime.version(), /^\d+\.\d+\.\d+$/)

function assertCppThrows(callback, expected) {
  let thrown
  try {
    callback()
  } catch (error) {
    thrown = error
  }
  assert.ok(thrown, 'Expected a C++ exception')
  const [, message] = runtime.getExceptionMessage(thrown)
  runtime.decrementExceptionRefcount(thrown)
  assert.match(message, expected)
}

runtime.registerNodeType('double', (encoded) => {
  const request = JSON.parse(encoded)
  const seed = request.state.channels.seed.value
  return JSON.stringify({
    writes: [{ channel: 'doubled', value: seed * 2 }],
  })
})
assert.deepEqual(JSON.parse(runtime.registeredNodeTypes()), ['double'])

const graph = new runtime.Graph(JSON.stringify({
  name: 'wasm-smoke',
  channels: {
    seed: { reducer: 'overwrite' },
    doubled: { reducer: 'overwrite' },
  },
  nodes: {
    double: { type: 'double' },
  },
  edges: [
    { from: '__start__', to: 'double' },
    { from: 'double', to: '__end__' },
  ],
}))

try {
  const result = JSON.parse(graph.run(JSON.stringify({
    threadId: 'wasm-smoke',
    input: { seed: 21 },
  })))
  assert.equal(result.output.channels.doubled.value, 42)
  assert.deepEqual(result.executionTrace, ['double'])
} finally {
  graph.delete()
  runtime.unregisterNodeType('double')
  assert.deepEqual(JSON.parse(runtime.registeredNodeTypes()), [])
}

runtime.registerNodeType('approval', (encoded) => {
  const request = JSON.parse(encoded)
  if (!request.context.resumeValue) {
    return JSON.stringify({
      interrupt: {
        reason: 'Approval required',
        value: { action: 'publish' },
      },
    })
  }
  return JSON.stringify({
    writes: [{
      channel: 'approved',
      value: request.context.resumeValue.approved,
    }],
  })
})

const approvalGraph = new runtime.Graph(JSON.stringify({
  name: 'wasm-hitl-smoke',
  channels: {
    approved: { reducer: 'overwrite', initial: false },
  },
  nodes: {
    approval: { type: 'approval' },
  },
  edges: [
    { from: '__start__', to: 'approval' },
    { from: 'approval', to: '__end__' },
  ],
}))

try {
  const paused = JSON.parse(approvalGraph.run(JSON.stringify({
    threadId: 'wasm-hitl',
    input: {},
  })))
  assert.equal(paused.interrupted, true)
  assert.equal(paused.interruptNode, 'approval')
  assertCppThrows(
    () => approvalGraph.resume('wasm-hitl', 'null'),
    /resume value must not be null/,
  )

  const resumed = JSON.parse(approvalGraph.resume(
    'wasm-hitl',
    JSON.stringify({ approved: true }),
  ))
  assert.equal(resumed.interrupted, false)
  assert.equal(resumed.output.channels.approved.value, true)
} finally {
  approvalGraph.delete()
  runtime.unregisterNodeType('approval')
}

console.log('NeoGraph WASM smoke passed')
